### PR TITLE
[REVIEW] Misc performance improvements for `scatter_by_map`

### PR DESCRIPTION
**Pull request by shwina**

### CODE REVIEW

File: python/cudf/cudf/_lib/column.pxd [hunk 1]
```diff
@@ -14,7 +14,18 @@ from cudf._lib.cpp.column.column_view cimport (
...
+        cdef int _offset
+        cdef int _size
+        cdef object _dtype
+        cdef object _base_children
+        cdef object _base_data
+        cdef object _base_mask
+        cdef object _children
+        cdef object _data
+        cdef object _mask
+        cdef object _null_count
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/column.pxd:27]: We should do the same for `table.pyx`

File: python/cudf/cudf/core/frame.py [hunk 2]
```diff
@@ -1066,12 +1066,16 @@ def _copy_categories(self, other, include_index=True):
                     offset=col.offset,
                 )
         if include_index:
-            from cudf.core.index import RangeIndex
+            from cudf.core.index import RangeIndex, CategoricalIndex
```
Comment (kkraus14) [inline python/cudf/cudf/core/frame.py:1069]: Can we move these imports outside of this function as well?

File: python/cudf/cudf/utils/dtypes.py [hunk 3]
```diff
@@ -80,39 +80,48 @@ def is_categorical_dtype(obj):
...
     if isinstance(
-        obj, (Index, Series, ColumnBase, pd.Index, pd.Series, np.ndarray)
+        obj,
+        (
+            cudf.Index,
+            cudf.Series,
+            cudf.core.column.ColumnBase,
+            pd.Index,
+            pd.Series,
+            np.ndarray,
```
Comment (kkraus14) [inline python/cudf/cudf/utils/dtypes.py:118]: Numpy will never have a categorical dtype, could likely short circuit this.

File: python/cudf/cudf/_lib/column.pxd (hunk 1 above)
Comment (shwina) [reply python/cudf/cudf/_lib/column.pxd:27]: So what makes this difficult/impossible right now is `RangeIndex`, which redefines `_data` as a property.

When `cdef` methods try to access the `._data` attributes, they don't evaluate the property, but rather the `cdef` attribute which breaks everything.

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/dataframe.py [hunk 4]
```diff
@@ -390,30 +391,24 @@ def __dir__(self):
...
+        # if an attribute already exists, set it.
+        try:
+            object.__getattribute__(self, key)
+            object.__setattr__(self, key, col)
+            return
+        except AttributeError:
+            pass
+
+        # if a column already exists, set it.
+        if key != "_data":
```
Comment (kkraus14) [inline python/cudf/cudf/core/dataframe.py:404]: When would `_data` throw `AttributeError`?

**Comment by kkraus14** [comment]
rerun tests

**Pull request merged**